src/Common/ReplicasReconnector.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include <Core/ServerSettings.h>
+#include <Core/BackgroundSchedulePool.h>
+#include <Interpreters/Context.h>
+#include <boost/noncopyable.hpp>
+#include <functional>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <utility>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NOT_INITIALIZED;
+    extern const int LOGICAL_ERROR;
+}
+
+namespace ServerSetting
+{
+    extern const ServerSettingsUInt64 bg_reconnect_mysql_dict_interval;
+}
+
+class ReplicasReconnector : private boost::noncopyable
+{
+public:
+    using Reconnector = std::function<bool(UInt64)>;
+    using ReconnectorsList = std::list<Reconnector>;
+
+    ReplicasReconnector(const ReplicasReconnector &) = delete;
+
+    ~ReplicasReconnector()
+    {
+        emergency_stop = true;
+        task_handle->deactivate();
+        instance_ptr(nullptr, true);
+    }
+
+    [[nodiscard]]
+    static std::unique_ptr<ReplicasReconnector> init(ContextPtr context)
+    {
+        std::unique_ptr<ReplicasReconnector> ret(new ReplicasReconnector(context));
+        instance_ptr(ret.get());
```

Comment:
Isn't it a memory leak or double-free? The `ret` object is being destroyed after leaving this function scope.

Reply:
`ReplicasReconnector::init()` is returning this `std::unique_ptr<ReplicasReconnector>` to `Server::main` local variable, so it undergoes RVO copy elision and nothing is destroyed at this point. And local variable in `Server::main` keeps this singleton alive during server's lifetime and destroys it when `Server::main` is finished - it's the whole point of "initializable singleton" - to keep its lifetime under control.